### PR TITLE
Handle request timeout correctly for HTTP

### PR DESCRIPTION
**Pull request by jrhee17**
Sorry about the delay and thanks for your patience 🙏 

**Motivation**
I've updated my thoughts in the issue: https://github.com/line/armeria/issues/3055#issuecomment-828560365

tldr; this PR basically attempts to suppress warning logs when a server-side initiated 408 request timeout message is sent to client side.

I wasn't able to reproduce this behavior with http2 (I'm guessing http2 would rely on RST_STREAM for this kind of behavior), so I've only handled this for http1.

**Modifications**
- Add a check for `Connection: close`, `status code: 408` in `client/Http1ResponseDecoder` when reading headers.

### CODE REVIEW

**Comment by trustin** [comment]
@jrhee17 Any progress on this?

**Comment by jrhee17** [comment]
> @jrhee17 Any progress on this?
back to work on this 🛠️

**Comment by trustin** [comment]
@jrhee17 Gentle ping :wink:

**Comment by jrhee17** [comment]
I think this PR is in a reviewable state - not sure why CI isn't running, but will resolve later

File: core/src/main/java/com/linecorp/armeria/client/Http1ResponseDecoder.java [hunk 1]
```diff
@@ -154,6 +154,7 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
             switch (state) {
                 case NEED_HEADERS:
                     if (msg instanceof HttpResponse) {
+
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/Http1ResponseDecoder.java:157]: Revert?
```suggestion
```

File: core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java [hunk 2]
```diff
@@ -1124,5 +1125,10 @@ public static String authorityHeader(String host, int port, int defaultPort) {
         }
     }
 
+    public static boolean isRequestTimeoutResponse(HttpResponse httpResponse) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java:1128]: A little bit of Javadoc would be nice, with a link to the related GH issue.

**Review by trustin — COMMENTED** [review]
Thanks!

File: core/src/main/java/com/linecorp/armeria/client/Http1ResponseDecoder.java [hunk 3]
```diff
@@ -166,6 +167,10 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                         }
 
                         final HttpResponseWrapper res = getResponse(resId);
+                        if (res == null && ArmeriaHttpUtil.isRequestTimeoutResponse((HttpResponse) msg)) {
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/Http1ResponseDecoder.java:170]:
```suggestion
                        if (res == null && ArmeriaHttpUtil.isRequestTimeoutResponse(nettyRes) {
```

File: core/src/test/java/com/linecorp/armeria/client/Http1ResponseDecoderTest.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.DefaultHttpResponse;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpVersion;
+
+public class Http1ResponseDecoderTest {
```
Comment (ikhoon) [inline core/src/test/java/com/linecorp/armeria/client/Http1ResponseDecoderTest.java:31]:
```suggestion
class Http1ResponseDecoderTest {
```

**Review by ikhoon — APPROVED** [review]
Finally, you nailed it. Thanks! @jrhee17

File: core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java [hunk 5]
```diff
@@ -1124,5 +1125,14 @@ public static String authorityHeader(String host, int port, int defaultPort) {
         }
     }
 
+    /**
+     * A 408 request timeout response can be received even without a request.
+     * More details can be found at https://github.com/line/armeria/issues/3055.
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java:1130]: 👍

File: core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java [hunk 6]
```diff
@@ -1124,5 +1125,14 @@ public static String authorityHeader(String host, int port, int defaultPort) {
         }
     }
 
+    /**
+     * A 408 request timeout response can be received even without a request.
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/internal/common/ArmeriaHttpUtil.java:1129]: nit: 
```suggestion
     * A 408 Request Timeout response can be received even without a request.
```

**Review by trustin — APPROVED** [review]
Thanks a lot, @jrhee17 !

**Pull request merged**